File: app/src/main/java/com/ja/assets/ui/activity/purchase/PurchaseApplyActivity.java
package com.ja.assets.ui.activity.purchase;

import android.content.Intent;
import android.view.KeyEvent;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.LinearLayoutCompat;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.fixed.u8.animation.RecyclerViewUtilKt;
import com.ja.assets.R;
import com.ja.assets.adapter.PurchaseApplyAdapter;
import com.ja.assets.databinding.ActivityPurchaseApplyBinding;
import com.ja.assets.model.AssetsInfo;
import com.ja.assets.model.PurchaseAssetsInfo;
import com.ja.assets.ui.base.BaseActivity;
import com.ja.assets.utils.view.BackDialog;

import java.util.ArrayList;
import java.util.List;


public class PurchaseApplyActivity extends BaseActivity {
    private ActivityPurchaseApplyBinding purchaseApplyBinding;
    private PurchaseApplyAdapter purchaseApplyAdapter;
    private List<PurchaseAssetsInfo> assetsInfoList;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_purchase_apply;
    }

    @Override
    protected void initView() {

        LinearLayoutCompat commonLeftLinearLayout = findViewById(R.id.commonLeftLinearLayout);
        commonLeftLinearLayout.setOnClickListener(v -> {
            finish();
        });
        TextView titleCenterText = findViewById(R.id.titleCenterText);
        titleCenterText.setText(R.string.purchaseApply);
        TextView titleRightText = findViewById(R.id.titleRightText);
        titleRightText.setText(R.string.addAssets);
        purchaseApplyBinding = (ActivityPurchaseApplyBinding) getViewDataBinding();


        titleRightText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(PurchaseApplyActivity.this, PurchaseNewAddActivity.class);
                intent.putExtra("assetsInfo", new PurchaseAssetsInfo("", "", "", 0, "", 0, "", "", "", ""));
                intent.putExtra("position", assetsInfoList.size());
                startActivityForResult(intent, 0);
            }
        });
        purchaseApplyBinding.submitAuditBtn.setOnClickListener(v -> toast("点击事件"));
    }

    @Override
    protected void initAdapter() {
        assetsInfoList = new ArrayList<>();
        setRecycler();
        RecyclerViewUtilKt recyclerViewUtilKt = new RecyclerViewUtilKt(this, purchaseApplyBinding.purchaseApplyRecycler);
        recyclerViewUtilKt.initRecyclerView();
        purchaseApplyAdapter = new PurchaseApplyAdapter(R.layout.item_purachse_apply_layout, assetsInfoList);
        recyclerViewUtilKt.setAdapter(purchaseApplyAdapter);
        purchaseApplyAdapter.setOnItemClickListener((adapter, view, position) -> {
            Intent intent = new Intent(this, PurchaseNewAddActivity.class);
            intent.putExtra("assetsInfo", assetsInfoList.get(position));
            intent.putExtra("position", position);
            startActivityForResult(intent, 1);
        });
        purchaseApplyAdapter.setOnItemChildClickListener((adapter, view, position) -> {
            switch (view.getId()) {
                case R.id.itemPurchaseApplyDelete:
                    assetsInfoList.remove(position);
                    purchaseApplyAdapter.notifyDataSetChanged();
                    setRecycler();
                    break;
                case R.id.itemPurchaseApplyEdit:
                    break;
                case R.id.itemPurchaseApplyUnAgree:
                    break;
                case R.id.itemPurchaseApplyAgree:
                    break;
                default:
                    break;
            }
        });
    }

    @Override
    protected void initData() {

    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (data == null) {
            return;
        }
        switch (requestCode) {
            case 0:
                PurchaseAssetsInfo assetsInfo = (PurchaseAssetsInfo) data.getSerializableExtra("assetsInfo");
                int position = data.getIntExtra("position", 0);
                assetsInfoList.add(assetsInfo);
                purchaseApplyAdapter.notifyDataSetChanged();
                setRecycler();
                break;
            case 1:
                PurchaseAssetsInfo assetsInfo1 = (PurchaseAssetsInfo) data.getSerializableExtra("assetsInfo");
                int position1 = data.getIntExtra("position", 0);
                assetsInfoList.set(position1, assetsInfo1);
                purchaseApplyAdapter.notifyDataSetChanged();
                setRecycler();
                break;
            default:
                break;
        }
    }

    private void setRecycler() {
        if (assetsInfoList.size() == 0) {
            purchaseApplyBinding.noAssetsLinear.setVisibility(View.VISIBLE);
            purchaseApplyBinding.purchaseApplyRecycler.setVisibility(View.GONE);
        } else {
            purchaseApplyBinding.noAssetsLinear.setVisibility(View.GONE);
            purchaseApplyBinding.purchaseApplyRecycler.setVisibility(View.VISIBLE);
        }
    }


    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.getRepeatCount() == 0) {
            new BackDialog(this).setListener(this::finish);
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }
}

File: app/src/main/java/com/ja/assets/ui/activity/purchase/PurchaseApplyActivity.kt
package com.ja.assets.ui.activity.purchase

import android.content.Intent
import android.view.KeyEvent
import android.view.View
import com.fixed.u8.animation.RecyclerViewUtilKt
import com.ja.assets.R
import com.ja.assets.adapter.PurchaseApplyAdapter
import com.ja.assets.databinding.ActivityPurchaseApplyBinding
import com.ja.assets.model.AssetsInfo
import com.ja.assets.ui.base.BaseActivity
import com.ja.assets.utils.view.BackDialog
import kotlinx.android.synthetic.main.activity_purchase_apply.*
import kotlinx.android.synthetic.main.common_title.*


class PurchaseApplyActivity : BaseActivity() {
    private var purchaseApplyBinding: ActivityPurchaseApplyBinding? = null
    private var purchaseApplyAdapter: PurchaseApplyAdapter? = null
    private var assetsInfoList: MutableList<AssetsInfo>? = null


    override fun getLayoutId(): Int = R.layout.activity_purchase_apply

    override fun initView() {
        commonLeftLinearLayout.setOnClickListener { finish() }
        titleCenterText.setText(R.string.purchaseApply)
        titleRightText.setText(R.string.addAssets)
        purchaseApplyBinding = viewDataBinding as ActivityPurchaseApplyBinding
    }

    override fun initAdapter() {
        assetsInfoList = mutableListOf()
        setRecycler()
        val recyclerViewUtilKt = RecyclerViewUtilKt(this, purchaseApplyRecycler)
        recyclerViewUtilKt.initRecyclerView()
        purchaseApplyAdapter = PurchaseApplyAdapter(R.layout.item_purachse_apply_layout, assetsInfoList!!)
        recyclerViewUtilKt.setAdapter(purchaseApplyAdapter!!)
        purchaseApplyAdapter?.setOnItemClickListener { adapter, view, position ->
            val intent = Intent(this, PurchaseNewAddActivity::class.java)
            intent.putExtra("assetsInfo", assetsInfoList!![position])
            intent.putExtra("position", position)
            startActivityForResult(intent, 1)
        }
        purchaseApplyAdapter?.setOnItemChildClickListener { adapter, view, position ->
            when (view.id) {
                R.id.itemPurchaseApplyDelete -> {
                    assetsInfoList!!.removeAt(position)
                    purchaseApplyAdapter?.notifyDataSetChanged()
                    setRecycler()
                }
                R.id.itemPurchaseApplyEdit -> {
                }
                R.id.itemPurchaseApplyUnAgree -> {
                }
                R.id.itemPurchaseApplyAgree -> {
                }
            }
        }
    }


    override fun initData() {
        titleRightText.setOnClickListener {
            val intent = Intent(this, PurchaseNewAddActivity::class.java)
            intent.putExtra("assetsInfo", AssetsInfo("", "", "", 0, "", 0, "", "", "", ""))
            intent.putExtra("position", assetsInfoList!!.size)
            startActivityForResult(intent, 0)
        }
        submitAuditBtn.setOnClickListener {
            toast("点击事件")
        }
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (data == null) {
            return
        }
        when (requestCode) {
            0 -> {
                val assetsInfo: AssetsInfo = data.getSerializableExtra("assetsInfo") as AssetsInfo
                val position: Int = data.getIntExtra("position", 0)
                assetsInfoList!!.add(assetsInfo)
                purchaseApplyAdapter?.notifyDataSetChanged()
                setRecycler()
            }
            1 -> {
                val assetsInfo: AssetsInfo = data.getSerializableExtra("assetsInfo") as AssetsInfo
                val position: Int = data.getIntExtra("position", 0)
                assetsInfoList!!.set(position, assetsInfo)
                purchaseApplyAdapter?.notifyDataSetChanged()
                setRecycler()
            }
        }
    }


    fun setRecycler() {
        if (assetsInfoList!!.size == 0) {
            noAssetsLinear.visibility = View.VISIBLE
            purchaseApplyRecycler.visibility = View.GONE
        } else {
            noAssetsLinear.visibility = View.GONE
            purchaseApplyRecycler.visibility = View.VISIBLE
        }
    }


    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK && event?.repeatCount == 0) {
            backDialog()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    private fun backDialog() {
        BackDialog(this).setListener(object : BackDialog.DateListener {
            override fun handleResult() {
                finish()
            }
        })
    }

}
